src/decision/assertion_list.cpp

```diff
@@ -0,0 +1,134 @@
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of assertion list
+ */
+
+#include "decision/assertion_list.h"
+
+namespace cvc5 {
+namespace decision {
+
+const char* toString(DecisionStatus s)
+{
+  switch (s)
+  {
+    case DecisionStatus::INACTIVE: return "INACTIVE";
+    case DecisionStatus::NO_DECISION: return "NO_DECISION";
+    case DecisionStatus::DECISION: return "DECISION";
+    case DecisionStatus::BACKTRACK: return "BACKTRACK";
+    default: return "?";
+  }
+}
+
+std::ostream& operator<<(std::ostream& out, DecisionStatus s)
+{
+  out << toString(s);
+  return out;
+}
+
+AssertionList::AssertionList(context::Context* ac,
+                             context::Context* ic,
+                             bool useDyn)
+    : d_assertions(ac),
+      d_assertionIndex(ic),
+      d_usingDynamic(useDyn),
+      d_dindex(ic)
+{
+}
+
+void AssertionList::presolve()
+{
+  Trace("jh-status") << "AssertionList::presolve" << std::endl;
+  d_assertionIndex = 0;
+  d_dlist.clear();
+  d_dindex = 0;
+}
+
+void AssertionList::addAssertion(TNode n) { d_assertions.push_back(n); }
+
+TNode AssertionList::getNextAssertion()
+{
+  size_t fromIndex;
+  if (d_usingDynamic)
+  {
+    // is a dynamic assertion ready?
+    fromIndex = d_dindex.get();
+    if (fromIndex < d_dlist.size())
+    {
+      d_dindex = d_dindex.get() + 1;
+      Trace("jh-status") << "Assertion " << d_dlist[fromIndex].getId()
+                         << " from dynamic list" << std::endl;
+      return d_dlist[fromIndex];
+    }
+  }
+  // check if dynamic assertions
+  fromIndex = d_assertionIndex.get();
+  Assert(fromIndex <= d_assertions.size());
+  if (fromIndex == d_assertions.size())
+  {
+    return Node::null();
+  }
+  // increment for the next iteration
+  d_assertionIndex = d_assertionIndex + 1;
+  Trace("jh-status") << "Assertion " << d_assertions[fromIndex].getId()
+                     << std::endl;
+  return d_assertions[fromIndex];
+}
+size_t AssertionList::size() const { return d_assertions.size(); }
+
+void AssertionList::notifyStatus(TNode n, DecisionStatus s)
+{
+  Trace("jh-status") << "Assertion status " << s << " for " << n.getId()
+                     << ", current " << d_dindex.get() << "/" << d_dlist.size()
+                     << std::endl;
+  if (!d_usingDynamic)
+  {
+    // not using dynamic ordering, return
+    return;
+  }
+  if (s == DecisionStatus::NO_DECISION)
+  {
+    // no decision does not impact the decision order
+    return;
+  }
+  std::vector<TNode>::iterator it =
+      std::find(d_dlist.begin(), d_dlist.end(), n);
+  if (s == DecisionStatus::DECISION)
+  {
+    if (it == d_dlist.end())
+    {
+      // if we just had status on an assertion and it didnt occur id dlist,
+      // then our index should have exhausted dlist
+      Assert(d_dindex.get() == d_dlist.size());
+      if (d_dindex.get() == d_dlist.size())
```

Comment:
Why the `if` if you already assert the condition?

Reply:
This is to make the code safe, I'd like it to fail in debug but not execute the branch otherwise (which can lead to segfaults).